### PR TITLE
feat(auto-cancel): server-side logic for auto canceling obsolete builds

### PR DESCRIPTION
**Pull request by ecrupper**
Depends on https://github.com/go-vela/types/pull/299

### CODE REVIEW

**Review by timhuynh94 — DISMISSED** [review]

File: api/build/auto_cancel.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+			// prepare the request to the worker
+			client := http.DefaultClient
+			client.Timeout = 30 * time.Second
+
+			// set the API endpoint path we send the request to
+			u := fmt.Sprintf("%s/api/v1/executors/%d/build/cancel", w.GetAddress(), executor.GetID())
+
+			req, err := http.NewRequestWithContext(context.Background(), "DELETE", u, nil)
+			if err != nil {
+				return err
```
[the remark this answers is not in the document]
Comment (ecrupper) [reply api/build/auto_cancel.go:140]: Do you mean like if there are multiple running builds to auto-cancel, and one of those attempts fails?

[the remark this answers is not in the document]
Comment (plyr4) [reply api/build/auto_cancel.go:140]: yeah, i was looking at the loop over executors

File: api/build/auto_cancel.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	"github.com/gin-gonic/gin"
+	"github.com/go-vela/server/database"
+	"github.com/go-vela/server/internal/token"
+	"github.com/go-vela/types/constants"
+	"github.com/go-vela/types/library"
+)
+
+// AutoCancel is a helper function that checks to see if any pending or running
+// builds for the repo can be replaced by the current build.
+func AutoCancel(c *gin.Context, b *library.Build, rBs []*library.Build, r *library.Repo, pending, running bool) error {
```
Comment (plyr4) [inline api/build/auto_cancel.go:23]:
```suggestion
func AutoCancel(c *gin.Context, b *library.Build, rBs []*library.Build, r *library.Repo, opts pipeline.CancelOptions) error {
```

File: api/build/auto_cancel.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+			continue
+		}
+
+		// ensure criteria is met before auto canceling
+		if (strings.EqualFold(rB.GetEvent(), constants.EventPush) &&
+			strings.EqualFold(b.GetEvent(), constants.EventPush) &&
+			strings.EqualFold(b.GetBranch(), rB.GetBranch())) ||
+			(strings.EqualFold(rB.GetEvent(), constants.EventPull) &&
+				strings.EqualFold(b.GetEventAction(), rB.GetEventAction()) &&
+				strings.EqualFold(b.GetHeadRef(), rB.GetHeadRef())) {
```
Comment (plyr4) [inline api/build/auto_cancel.go:37]: should we add a helper func? or comments? this `if` is gnarly and my eyes glaze over it :(

File: api/build/auto_cancel.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ecrupper) [reply api/build/auto_cancel.go:140]: Only one build will match, so there should only be one attempt at cancellation. 

But we do loop over pending/running builds: https://github.com/go-vela/server/pull/911/files#diff-2e48a4018b70aac91235c0a2f5ec50d7ae186de39e3bd603d7e0fcfe6fd5ab07R25-R65.

I was unsure as to whether or not to immediately error out upon failing one of those (provided they match all the other criteria). What do you think? If this policy was implemented, there really should only be one build at a time that gets the boot. Each new build will supersede the older build, which superseded a previous build, and so on.

All's that to say, even though there are a couple loops here, we're really just performing one operation overall — I think... 😅

**Review by wass3rw3rk — COMMENTED** [review]
It seems like the current implementation has the potential to take some time before the current build makes it onto the queue dependent on how many builds need to be canceled and how many containers each build spun up.

It waits until all builds are canceled before pushing the current build to the queue. Should we send current build to queue and, in best effort, send off cancellation attempts to the builds that we need to cancel?

**Review by wass3rw3rk — COMMENTED** [review]
also, if it fails canceling any one of the builds it will not attempt to cancel any of the other builds it needs to cancel, i believe? i could be wrong - but if not, i think we want to continue attempting to cancel the other builds.

**Review by wass3rw3rk — APPROVED** [review]

**Review by timhuynh94 — APPROVED** [review]

**Pull request merged**